[Kernel] AQ AZP 4/4: Integrate asymmetric quantization to linear method

ProExpertProg opened the pull request:
Add end-to-end support for (dynamic) asymmetric activation quantization by invoking AZP quant and scaled_mm kernels in linear method. Both dynamic asymmetric AQ and static asymmetric AQ implemented.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

ProExpertProg committed:
Added asymmetric integration to linear layers

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py:
```diff
@@ -83,11 +98,18 @@ def create_weights(self, layer: torch.nn.Module,
                 output_partition_sizes, **layer_kwargs)
             layer.register_parameter("input_scale", input_scale)
 
+            if not self.input_symmetric:
```
Why can't this be supported?

ProExpertProg replied:
It could be, it's just untested because I was told we don't care about that case

mgoin replied:
Isn't this what this PR works on supporting?

tlrmchlsmth replied:
This PR adds support for dynamic per-token asymmetric quantized activations. I think this is more likely to be used than the static, per-tensor, asymmetric case, but do we have all the pieces implemented for that?

ProExpertProg replied:
We're only missing the integration piece for static asymmetric. ops & kernels are all tested supported and working

ProExpertProg replied:
Replaced with warning

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py:
```diff
@@ -44,8 +46,21 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+            layer.input_zero_point = None
+
+        if not self.input_symmetric:
+            layer.azp_adj = layer.weight.sum(dim=0,
```
Can you explain more what `azp_adj` is?

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/quantization/utils/w8a8_utils.py:
```diff
@@ -194,13 +194,28 @@ def apply_int8_linear(
+                                               symmetric=symmetric)
+
+    if x_zp is not None:
+        return ops.cutlass_scaled_mm_azp(x_q,
```
Do you think it would be better to have `ops.cutlass_scaled_mm` with optional azp parms instead of branching here?

ProExpertProg replied:
I just didn't want the `cutlass_scaled_mm` to blow up in complexity. Also, in the dynamic case, we don't pass `azp`  because it's calculated, so we'd need an extra parameter to distinguish the symmetric and asymmetric cases

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py:
```diff
@@ -44,8 +46,21 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         if self.is_static_input_scheme:
             layer.input_scale = Parameter(layer.input_scale.max(),
                                           requires_grad=False)
+            if not self.input_symmetric:
```
I dont think this is needed

ProExpertProg replied:
Hmm interesting, yeah not sure how to adapt this for asymmetric, would probably have to reconstruct the ranges here

ProExpertProg replied:
This case is actually not necessary if we don't support the static per-tensor case

ProExpertProg commented:
/ready

ProExpertProg committed:
PR comments

mgoin commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py:
```diff
@@ -46,10 +48,38 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+
+        if not self.input_symmetric:
+            # azp_adj is the AZP adjustment term, used to account for weights.
+            # For more details, see csrc/quantization/cutlass_w8a8/Epilogues.md
```
nit: would be best to use a github link here in case this file moves

ProExpertProg committed:
PR comments

ProExpertProg commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py:
```diff
@@ -46,10 +48,38 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+                layer.input_scale = Parameter(layer.input_scale.max(),
+                                              requires_grad=False)
+            else:
+                raise NotImplementedError(
+                    "static input asymmetric quantization not supported yet")
```
The code after this exception is completely untested, but I can remove the code, or add a comment

ProExpertProg committed:
Warning instead of exception

ProExpertProg committed:
Added epilogues link

ProExpertProg commented:
Replaced the exception with a warning for untested code paths

ProExpertProg commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py:
```diff
@@ -90,11 +128,19 @@ def create_weights(self, layer: torch.nn.Module,
+                # Static asymmetric quantization has not been tested yet
+                logger.warning(
+                    "Static asymmetric quantization currently untested")
+                input_zero_point = Parameter(torch.zeros(1, dtype=torch.int32))
```
@dsikka should I use a BasevLLMParameter here?

tlrmchlsmth commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py:
```diff
@@ -46,10 +51,43 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+                # Static asymmetric quantization has not been tested yet.
+                # Kernel and ops support exists and is tested, it's just the
+                # following integration code that is untested.
+                logger.warning(
+                    "Static asymmetric quantization currently untested")
```
Let's beef up this warning message a bit:
```suggestion
                logger.warning(
                    "Support for models with static asymmetric quantized activations has been implemented but is untested.")
```

tlrmchlsmth replied:
Hmm, do we have a nice way to only log once? This is going to cause a ton of spam if somebody actually runs into this

tlrmchlsmth replied:
Try this:
```suggestion
                vllm.utils.print_warning_once(
	                    "Support for models with static asymmetric quantized activations has been implemented but is untested.")
```

tlrmchlsmth commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py:
```diff
@@ -90,11 +128,19 @@ def create_weights(self, layer: torch.nn.Module,
             layer.register_parameter("input_scale", input_scale)
 
+            if not self.input_symmetric:
+                # Static asymmetric quantization has not been tested yet
+                logger.warning(
+                    "Static asymmetric quantization currently untested")
```
IMO we don't need both warnings.

ProExpertProg committed:
Improve warning

Co-authored-by: Tyler Michael Smith <tyler@neuralmagic.com>

ProExpertProg committed:
Remove 2nd warning

ProExpertProg committed:
BasevLLMParameter

ProExpertProg committed:
print warning once

dsikka reviewed (commented):
Do we have example models where this has been tested?
If yes, can they be added to this test: https://github.com/vllm-project/vllm/blob/main/tests/weight_loading

dsikka commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py:
```diff
@@ -46,10 +52,44 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+
+                scale = (range_max - range_min) / (int8_traits.max -
+                                                   int8_traits.min)
+                layer.input_scale = Parameter(scale, requires_grad=False)
```
I think we should add an accuracy test to make sure this works as expected

ProExpertProg replied:
Where should we add the accuracy test?

dsikka replied:
this is now complete with the compressed tensors quantization tests - thanks!

ProExpertProg committed:
Add dynamic asymmetric models to testing

ProExpertProg commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py:
```diff
@@ -46,10 +52,44 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
             layer.input_scale = None
+            layer.input_zero_point = None
+
+        if not self.input_symmetric:
```
Yes that's correct, this term only depends on the weights, not azp or scales. Added a comment

ProExpertProg committed:
More PR comments

ProExpertProg committed:
Add lm-eval

dsikka committed:
fix bug

dsikka committed:
update dtype for zp; add static asym case

dsikka commented:
@ProExpertProg  I fixed the quantization tests that were previously failing. I've also updated the dtype for zero points for the static case and added a static test model to quantization tests.

The model currently fails due to a dtype error

ProExpertProg committed:
Fixed static azp loading case

ProExpertProg committed:
Added asymmetric model to lm-eval small models

ProExpertProg commented:
@dsikka thanks for the help! I fixed the failing test case.

I also removed the warning for the static asymmetric case because the code is no longer untested.

dsikka reviewed (approved):
LGTM pending all quantization tests pass

tlrmchlsmth reviewed (approved):
Thanks for getting everything tested!

ProExpertProg committed:
Fixed lm-eval scores

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into luka/aq-azp-linear

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into luka/aq-azp-linear

mgoin commented on .buildkite/lm-eval-harness/test_lm_eval_correctness.py:
```diff
@@ -49,10 +49,15 @@ def test_lm_eval_correctness():
             print(f'{task["name"]} | {metric["name"]}: '
                   f'ground_truth={ground_truth} | measured={measured_value}')
-            assert numpy.isclose(ground_truth, measured_value, rtol=RTOL)
+            success = success and numpy.isclose(
+                ground_truth, measured_value, rtol=RTOL)
+
+    # Assert at the end, print all scores even on failure for debugging.
+    assert success
```
The issue with this is while you do get to see all the scores, you don't know quickly which one was out of tolerance. You need to check the differences yourself?

ProExpertProg replied:
Ok what if I add the success as a column in the results? It's really annoying to not see all scores on failure
